In [218]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Project-2: Locality Sensitive Hashing

In [219]:
import random
import os
import numpy as np

from random import randrange
from shutil import copyfile

import binascii

## Execute the follwing two cells to generate your data sample

In [220]:
n_1=0
n_2=0
n_3=0

In [221]:
MY_ID = '692'
# MY_ID should be string
# Example MY_ID='819'

task_dict = {0:'taska',1:'taskb',2:'taskc',3:'taskd',4:'taske'}
id_dict = {0:n_1,1:n_2,2:n_3}

try:
    n_1 = int(MY_ID[-1]) % 5
    n_2 = int(MY_ID[-2]) % 5
    n_3 = int(MY_ID[-3]) % 5
    
    while n_1 == n_2 or n_1 == n_3:
        n_1 = (n_1 + 1) % 5
    
    while n_1 == n_2 or n_2 == n_3:
        n_2 = (n_2 + 1) % 5
        
except Exception as e:
    print('Please enter a valid ID...')

In [222]:
data_path = '/content/drive/My Drive/XXXXXX/corpus-20090418' 
# Edit this path if the data directory is not in the current directory

try:
    os.makedirs('Data_Sample')
    os.makedirs('Original_Sample')
except Exception as e:
    pass

for i in range(3):
    task = task_dict[id_dict[i]]
    
    for file in os.listdir(data_path):
        if task in file:
            if 'orig' in file:
                copyfile(data_path + '/' + file, 'Original_Sample/' + file)
            else:
                copyfile(data_path + '/' + file, 'Data_Sample/' + file)

###Finding the number of docs created in Data_Sample and Original_Sample folders

In [223]:
p=0
q=0
for filename in os.listdir("/content/Data_Sample"):
    answerdoc=os.path.join("/content/Data_Sample",filename)
    p+=1
for filename in os.listdir("/content/Original_Sample"):
    answerdoc=os.path.join("/content/Original_Sample",filename)
    q+=1
print("Number of documents created in Data_Sample folder : ", str(p))
print("Number of documents created in Original_Sample folder : ", str(q))

Number of documents created in Data_Sample folder :  19
Number of documents created in Original_Sample folder :  1


Your dataset for this project will be in <b>Data_Sample</b>


Your query choices will be in <b>Original_Sample</b> directory


You have to use any one original Wikipedia article from <b>Original_Sample</b> for <b>Fact Check</b> steps

### STEP - 1: Shingling (20 points)

In [224]:
# Type your code here... 
# Create necessary number of cells below this cell
# k = 3

# NOTE: No complex text processing is required
# convert just upper case characters to lower case


In [225]:
samplefolder_3shingles=set()
samplefolder_4shingles=set()
samplefolder_5shingles=set()

In [226]:
def task1(folder_path):
  for filename in os.listdir(folder_path):
    answerdoc=os.path.join(folder_path,filename)
    f = open(answerdoc, "r",encoding='utf-8',errors='ignore')
    answer=f.read()
    tokens=answer.split()
  # Creating unique 3-shingles
    k=3
    for i in range(0,len(tokens)-k+1):
      samplefolder_3shingles.add(str(tokens[i]+" "+tokens[i+1]+" "+tokens[i+2]).lower())
  # Creating unique 4-shingles
    k=4
    for i in range(0,len(tokens)-k+1):
      samplefolder_4shingles.add(str(tokens[i]+" "+tokens[i+1]+" "+tokens[i+2]+" "+tokens[i+3]).lower())
  # Creating unique 5-shingles
    k=5
    for i in range(0,len(tokens)-k+1):
      samplefolder_5shingles.add(str(tokens[i]+" "+tokens[i+1]+" "+tokens[i+2]+" "+tokens[i+3]+" "+tokens[i+4]).lower())

In [227]:
task1('/content/Data_Sample')
task1('/content/Original_Sample')

In [228]:
print('Number of unique 3-shingles:' ,len(samplefolder_3shingles))
print('Number of unique 4-shingles:' ,len(samplefolder_4shingles))
print('Number of unique 5-shingles:' ,len(samplefolder_5shingles))

Number of unique 3-shingles: 3133
Number of unique 4-shingles: 3347
Number of unique 5-shingles: 3439


Report results (number of unique k-shingles) for k={3,4,5} below:
1. k=3: 3133
2. k=4: 3347
3. k=5: 3439

In [229]:
# Type your code to get the 5-shingle index here


In [230]:
# Creating 5-shingle index. In the data sample folder, for each doc, 5-shingles are created and stored in eachdoc_5shingles list. All these lists are added into
# fiveShingles_index dictionary with key as the doc name and value as its corresponding 5-shingles ascii values.
# So, fiveShingles_index length is nothing but number of documents in the sample folder.
# The value for each key in fiveShingles_index dictonary is 5shingles and length of that value is the number of unique shingles in that particular document.
def generate_shingles(filepath,k):
  fiveShingles_index={}
  for filename in os.listdir(filepath):
    answerdoc=os.path.join(filepath,filename)
    f = open(answerdoc, "r",encoding='utf-8',errors='ignore')
    answer=f.read()
    tokens=answer.split()
    eachdoc_5shingles=[]
    for i in range(0,len(tokens)-k+1):
      c=str(tokens[i]+" "+tokens[i+1]+" "+tokens[i+2]+" "+tokens[i+3]+" "+tokens[i+4]).lower()
      c=binascii.crc32(bytes(c,'utf-8'))
      if c not in eachdoc_5shingles:
        eachdoc_5shingles.append(c)
    fiveShingles_index[filename]=eachdoc_5shingles
  return fiveShingles_index

In [231]:
data_Sample_5shingleindex=generate_shingles('/content/Data_Sample',5)
Original_Sample_5shingleindex=generate_shingles('/content/Original_Sample',5)

In [1]:
# Printing 5 shingles ascii values for each doc in Data_Sample folder
#for k, v in data_Sample_5shingleindex.items():
  #print(k,v)

In [2]:
# Printing 5 shingles ascii values for each doc in Original_Sample folder
#for k, v in Original_Sample_5shingleindex.items():
  #print(k,v)

In [234]:
#Getting the count of unique 5 shingles from Data_sample folder and Orig_Sample folder
all_unique_5shingles_index=set()
for k, v in data_Sample_5shingleindex.items():
  for i in v:
   all_unique_5shingles_index.add(i)
for k, v in Original_Sample_5shingleindex.items():
  for i in v:
   all_unique_5shingles_index.add(i)

In [235]:
len(all_unique_5shingles_index)

3439

In [236]:
# Just a cross check for my understanding
# Here, I am checking the total number of unique shingles from each doc. This is equal to 4142. This is different than the number 
# of unique shingles in all the documents, which is 3439 as calculated in one of the above cells. This means some shingles are
# common even across the different documents. Hence 3439(count of unique shingles from all docs) < 4142( sum of unique shingles from each doc)
t=0
for k, v in data_Sample_5shingleindex.items():
  #print(len(v))
  t+=len(v)
print(t)

4142


### STEP - 2: Min-Hashing (40 points)

In [237]:
# ***************************************************NEW***********************************************************
# Generate Hash functions - 
    # We use (ax + b) mod N formula to permute shingle index
    # where a,b are random numbers, N total index size, and x is the index
# We need to do L permutations - In other words we need to have L permutations (lists) of new indexes
# Following function takes total index size N and L as arguments
    # And returns L new lists of size N
    
def get_hash_functions(N,L):
    hash_functions = []
    
    for itr in range(L):
        a=randrange(1,400)
        b=randrange(1,400)
        
        new_hash_function = []
        for i in range(N):
            new_hash_function.append((a * i + b) % N)
        
        hash_functions.append(new_hash_function)
    return hash_functions
        
# test
# hash_functions = get_hash_functions(3439,50)

In [238]:
L= [50,100,200,500,1000]

In [239]:
# Type your code here to generate all L hash functions
# Generate hash functions only for shingle index created for k=5
#Generating a
h_f=[]
for i in L:
  hash_functions=get_hash_functions(len(all_unique_5shingles_index),i)
  h_f.append(hash_functions)

In [240]:
print("Number of L's:",str(len(h_f)))
print("Number of hash functions in each of the 5 L's:")
for i in range(0,len(h_f)):
  print(len(h_f[i]))

Number of L's: 5
Number of hash functions in each of the 5 L's:
50
100
200
500
1000


In [189]:
# Type your code here to get the final signature matrix S here


In [241]:
def generate_signaturematrix(docsshingles_index,allshingles_index,hash_functions,L):
  all=[]
  #docs=[]
  for k,v in docsshingles_index.items():
    middle=[np.inf for r in [None]*L]
    #for i in range(0,3439):
    for m in allshingles_index:
      if (m in v):
        i=list(allshingles_index).index(m)
        for j in range(0,len(hash_functions)):
          if hash_functions[j][i]<middle[j]:
            middle[j]=hash_functions[j][i]
            #print(middle) #use this print to check
    #docs.append(k)
    all.append(middle)
  return np.array(all).T

In [242]:
# Reporting signature matrices for each L hash fucntions where L=[50,100,200,500,1000]
all_orig_sign_matrices=[]
all_data_sign_matrices=[]
for z in range(0,len(h_f)):
  orig_sample_signmatrix=generate_signaturematrix(Original_Sample_5shingleindex,all_unique_5shingles_index,h_f[z],len(h_f[z]))
  data_sample_signmatrix=generate_signaturematrix(data_Sample_5shingleindex,all_unique_5shingles_index,h_f[z],len(h_f[z]))
  all_orig_sign_matrices.append(orig_sample_signmatrix)
  all_data_sign_matrices.append(data_sample_signmatrix)

In [243]:
for i in range(0,len(all_data_sign_matrices)):
  # all_data_sign_matrices is a list of signature matrices for L=[50,100,200,500,1000] respectively
  print("Data Sample folder signature matrices size for L =",str(len(all_data_sign_matrices[i])),":",str(all_data_sign_matrices[i].shape))
  # all_orig_sign_matrices is a list of signature matrices for L=[50,100,200,500,1000] respectively
  print("Original Sample folder signature matrices size for L =",str(len(all_orig_sign_matrices[i])),":",str(all_orig_sign_matrices[i].shape))

Data Sample folder signature matrices size for L = 50 : (50, 19)
Original Sample folder signature matrices size for L = 50 : (50, 1)
Data Sample folder signature matrices size for L = 100 : (100, 19)
Original Sample folder signature matrices size for L = 100 : (100, 1)
Data Sample folder signature matrices size for L = 200 : (200, 19)
Original Sample folder signature matrices size for L = 200 : (200, 1)
Data Sample folder signature matrices size for L = 500 : (500, 19)
Original Sample folder signature matrices size for L = 500 : (500, 1)
Data Sample folder signature matrices size for L = 1000 : (1000, 19)
Original Sample folder signature matrices size for L = 1000 : (1000, 1)


In [3]:
# Printing signature matrices for all L=50 for data and original sample files. As its lenghty to print, 
# printing only for L=50, To get for other L's
# the index can be cahnegd
#print(all_data_sign_matrices[0])
#print(all_orig_sign_matrices[0])

####Calculcating Jaccard Similarity

In [248]:
# Function to get the jaccard similarity. I used this code avaialble on the internet.
def jaccard(list1, list2):
    intersection = len(list(set(list1).intersection(list2)))
    union = (len(list1) + len(list2)) - intersection
    return round(float(intersection) / union,4)

In [249]:
def generate_jaccard_similarity(orig_signature_matrix,data_signature_matrix):
  all_sim=[]
  for i in range(0,5):
    sim=[]
    blue=orig_signature_matrix[i][:,0].tolist()
    for j in range(0,p):
      green=data_signature_matrix[i][:,j].tolist()
      sim.append(jaccard(blue,green))
    all_sim.append(sim)
  jsim_all=[]
  for m in range(0,len(all_sim)):
    j_sim={}
    for i in range(0,p):
      j_sim[doc_names[i]]=all_sim[m][i]
    jsim_all.append(j_sim)
  return jsim_all

In [250]:
#Here, jaccard similarity for all 19 docs with respect to original is calculated.
# This is performed across asll 5 L's
# The resulting all_sim is a list of 5 elements representing L=[50,100,200,500,1000] resp
# Each element in jsim_all is a dictionary with keys as doc names and values as its jaccard similarities with original
all_sim=[]
for i in range(0,len(L)):
  sim=[]
  blue=all_orig_sign_matrices[i][:,0].tolist()
  for j in range(0,p):
    green=all_data_sign_matrices[i][:,j].tolist()
    sim.append(jaccard(blue,green))
  all_sim.append(sim)

In [251]:
doc_names=[]
for k in data_Sample_5shingleindex.keys():
  doc_names.append(k)

In [252]:
jsim_all=[]
for m in range(0,len(all_sim)):
  j_sim={}
  for i in range(0,p):
    j_sim[doc_names[i]]=all_sim[m][i]
  jsim_all.append(j_sim)

In [253]:
#Each element in jsim_all has jaccard similartoes for each of L=[50,100,200,500,1000]. Here, I printed for L=1000
# Fact Check
jsim_all[4]

{'g0pA_taska.txt': 0.0288,
 'g0pB_taska.txt': 0.0277,
 'g0pC_taska.txt': 0.0293,
 'g0pD_taska.txt': 0.0299,
 'g0pE_taska.txt': 0.0293,
 'g1pA_taska.txt': 0.0293,
 'g1pB_taska.txt': 0.0299,
 'g1pD_taska.txt': 0.0299,
 'g2pA_taska.txt': 0.0283,
 'g2pB_taska.txt': 0.0293,
 'g2pC_taska.txt': 0.0293,
 'g2pE_taska.txt': 0.0262,
 'g3pA_taska.txt': 0.0299,
 'g3pB_taska.txt': 0.0299,
 'g3pC_taska.txt': 0.0293,
 'g4pB_taska.txt': 0.0293,
 'g4pC_taska.txt': 0.0283,
 'g4pD_taska.txt': 0.0299,
 'g4pE_taska.txt': 0.0293}

**Entire code for creating 5 shingles index, generating signature matrix for each of the L=[50,100,200,500,1000] and calculating jaccard similarities-fact check**

In [254]:
# Type your code here to do the fact check 
#      with any one query document in the 'Original_Sample' directory

original = 'Original_Sample/orig_taskc.txt'

# STEP-1: Generate 5-shingles 
    # (if any shingles are not present in your shingle index, simply ignore them)
data_Sample_5shingleindex=generate_shingles('/content/Data_Sample',5)
Original_Sample_5shingleindex=generate_shingles('/content/Original_Sample',5)
    
all_unique_5shingles_index=set()
for k, v in data_Sample_5shingleindex.items():
  for i in v:
   all_unique_5shingles_index.add(i)
for k, v in Original_Sample_5shingleindex.items():
  for i in v:
   all_unique_5shingles_index.add(i)    
    
# STEP-2: Generate signature vector from L hash functions
L=[50,100,200,500,1000]

h_f=[]
for i in L:
  hash_functions=get_hash_functions(len(all_unique_5shingles_index),i)
  h_f.append(hash_functions)

all_orig_sign_matrices=[]
all_data_sign_matrices=[]
for z in range(0,len(h_f)):
  orig_sample_signmatrix=generate_signaturematrix(Original_Sample_5shingleindex,all_unique_5shingles_index,h_f[z],len(h_f[z]))
  data_sample_signmatrix=generate_signaturematrix(data_Sample_5shingleindex,all_unique_5shingles_index,h_f[z],len(h_f[z]))
  all_orig_sign_matrices.append(orig_sample_signmatrix)
  all_data_sign_matrices.append(data_sample_signmatrix)

doc_names=[]
for k in data_Sample_5shingleindex.keys():
  doc_names.append(k)

# STEP-3: Calculate Jaccard similarity of signature vector of orginal doc.
    # and all other documents 
jaccard_similarities=generate_jaccard_similarity(all_orig_sign_matrices,all_data_sign_matrices)
print("************************  Jaccard Similarities for L=[50,100,200,500,1000] without any threshold*******************************************")
print(jaccard_similarities)
t = 0.20
d=[]
for e in range(0,len(jaccard_similarities)):
  d1=[]
  for k,v in jaccard_similarities[e].items():
    if v<t:
      d1.append(k)
  d.append(d1)
s=0
while s<len(d):
  #print(s)
  for i in d[s]:
    jaccard_similarities[s].pop(i)
  s+=1
print("************************* Jaccard Similarities for L=[50,100,200,500,1000] greater than t=0.2 **************************")
print(jaccard_similarities) 

************************  Jaccard Similarities for L=[50,100,200,500,1000] without any threshold*******************************************
[{'g3pA_taska.txt': 0.1765, 'g0pB_taska.txt': 0.1905, 'g4pB_taska.txt': 0.1905, 'g0pD_taska.txt': 0.1765, 'g1pA_taska.txt': 0.1494, 'g1pD_taska.txt': 0.1905, 'g0pC_taska.txt': 0.1905, 'g4pC_taska.txt': 0.2195, 'g0pE_taska.txt': 0.2346, 'g2pC_taska.txt': 0.1905, 'g3pC_taska.txt': 0.1765, 'g2pB_taska.txt': 0.1765, 'g0pA_taska.txt': 0.2048, 'g4pD_taska.txt': 0.1628, 'g2pE_taska.txt': 0.1628, 'g1pB_taska.txt': 0.2195, 'g2pA_taska.txt': 0.2048, 'g4pE_taska.txt': 0.2048, 'g3pB_taska.txt': 0.1905}, {'g3pA_taska.txt': 0.1429, 'g0pB_taska.txt': 0.1429, 'g4pB_taska.txt': 0.1494, 'g0pD_taska.txt': 0.1494, 'g1pA_taska.txt': 0.1494, 'g1pD_taska.txt': 0.1429, 'g0pC_taska.txt': 0.1299, 'g4pC_taska.txt': 0.1696, 'g0pE_taska.txt': 0.1561, 'g2pC_taska.txt': 0.1299, 'g3pC_taska.txt': 0.1429, 'g2pB_taska.txt': 0.1364, 'g0pA_taska.txt': 0.1364, 'g4pD_taska.txt': 0.1494

***************************************************NEW***********************************************************
For each L = {50,100,200,500,1000}, report all documents (file_names) below that have Jaccard similarity > t=0.85
Sort the documents in decreasing order of the Jaccard similarity


In [39]:
#This is reported above. I do not get any docs with t>0.2 except for L=50

### STEP - 3: LSH (30 points)

In [269]:
#Generating signature matrix for L=1000
data_Sample_5shingleindex=generate_shingles('/content/Data_Sample',5)
Original_Sample_5shingleindex=generate_shingles('/content/Original_Sample',5)

all_unique_5shingles_index=set()
for k, v in data_Sample_5shingleindex.items():
  for i in v:
   all_unique_5shingles_index.add(i)
for k, v in Original_Sample_5shingleindex.items():
  for i in v:
   all_unique_5shingles_index.add(i)  

L=1000
hash_functions=get_hash_functions(len(all_unique_5shingles_index),L)
orig_sample_1000signmatrix=generate_signaturematrix(Original_Sample_5shingleindex,all_unique_5shingles_index,hash_functions,L)
data_sample_1000signmatrix=generate_signaturematrix(data_Sample_5shingleindex,all_unique_5shingles_index,hash_functions,L)

In [270]:
print("Size of Data Sample Signature matrix:",str(data_sample_1000signmatrix.shape))
print(data_sample_1000signmatrix)
print("Size of Original Sample Signature Matrix:",str(orig_sample_1000signmatrix.shape))
#print(orig_sample_1000signmatrix)

Size of Data Sample Signature matrix: (1000, 19)
[[56 18 18 ... 18 18 56]
 [11  0 18 ... 25 14  3]
 [57 12  5 ...  0 19  8]
 ...
 [ 5  5  5 ...  5 43  5]
 [27 19 11 ... 14  7 31]
 [30 11  4 ...  0  8  1]]
Size of Original Sample Signature Matrix: (1000, 1)


In [271]:
# Type your code here to hash signature matrix into B buckets
# Use the technique to split the signature matrix into b bands of r rows
# Convert only the signature matrix generated with L=1000

b = 50
r = 20
B = 199


In [272]:
def generate_bands_hash(signaturematrix,bands,rowsinband,buckets):
  allbands=[]
  for z in range(0,bands): # 50 is number of bands
    l=0
    h=rowsinband
    x=signaturematrix[l:h]
    band=[]
    for i in range(0,len(signaturematrix[0])):
      zoo=random.sample(range(100),rowsinband)
      igloo=0 
      for j in range(0,rowsinband):
        igloo+=x[:,i][j]*zoo[j] 
      band.append(igloo % buckets)
    allbands.append(band)
    l=h
    h=h+rowsinband
  return np.array(allbands)

Generating Candidate Documents, False Positives, False Ngeatives
Now, I have 50 bands for 19 docs and 50 bands for 1 original doc.
Candidate docs are identified using the following criteria.
For doc1, band1(doc1) is compared with band1(orig).
In the 50 bands, if atleast one band value is common, which means they fall into same bucket. In this case, doc1 is a candidate document. This process is repeated over 19 docs comparing with original for all corresponding 50 bands.

In [273]:
def generate_cand_fp_fn(orig_hash_sig_matrix,data_hash_sig_matrix,t):
  hava={}
  candidate_documents=[]
  o=orig_hash_sig_matrix[:,0].tolist()
  for i in range(0,p):
    d=data_hash_sig_matrix[:,i].tolist()
    count=0
    for j in range(0,b):
      if o[j]==d[j]:
        count+=1
    if count>=1:
      hava[doc_names[i]]=count
    #print(doc_names[i])
      candidate_documents.append(doc_names[i])
  print("************************* Candidate Documents ********************************")
  print(candidate_documents)
  print("                         ")
  print("************* Candidate documents and how many buckets in common with the original document ************")
  print(hava)
  print("                         ")
  plus={}
  j=[]
  o=orig_hash_sig_matrix[:,0].tolist()
  for i in range(0,p):
    d=data_hash_sig_matrix[:,i].tolist()
    j.append(jaccard(o,d))
    plus[doc_names[i]]=jaccard(o,d)
# Reporting jaccard similarites for candidate docs
  print("************* Jaccard Similarities of Candidate Documents *****************")
  for k, v in plus.items():
    if k in candidate_documents:
      print(k,v)
  print("                         ")
  false_positives={}
  for c in candidate_documents:
    if plus[c]<=t:
      false_positives[c]=plus[c]
  print("                                              ")
  print("****************** All documents with Jaccard Similarity >t=0.1************************")
  alldocgreat_jt=[]
  for k, v in plus.items():
    if v>t:
     alldocgreat_jt.append(k)
  print(alldocgreat_jt)
  print("                                     ")
  print("Count of False Positives:", len(false_positives))
  print("********* False positive documents and their corresponding Jaccard Similarity *************")
  print(false_positives)
  print("                         ")
  false_negatives={}
  for k,v in plus.items():
    if v>t and k not in candidate_documents:
      false_negatives[k]=v
  print("Count of False Negatives:", len(false_negatives))
  print("************ False negative documents and their corresponding Jaccard Similarity *************")
  print(false_negatives)

In [274]:
# Type your code here to do generate candidate documents
# Follow all steps from STEP - 2 fact check (except the Jaccard similarity part)

# STEP - 1: Split your original document signature vector into b bands of r rows

# STEP - 2: Hash using the same hash functions created for 
    # signature matrix hashing (in the previous cell)
orig_bands_hash=generate_bands_hash(orig_sample_1000signmatrix,b,r,B)
orig_bands_hash.shape

(50, 1)

In [275]:
# Type your code here to do the fact check
# Calculate Jaccard similarity of the roiginal document with only 
    # candidate documents

Report all documents (file_names) below that have Jaccard similarity > t=0.85
Sort the documents in decreasing order of the Jaccard similarity

Report the list of false positives and false negatives below

I reported **candidaite docs**, **all docs with jaccard similarity >t=0.1** and **false positives**, **false negatives** in the below cell.
As I have only 19 docs, none of the docs have above 0.2. The max I am gettinf is 0.19. So,I am checking with t=0.1

In [276]:
b = 50
r = 20
B = 199
print("                                                                                                             ")
print("******************** Results for b =", str(b),"r =", str(r), "B =", str(B),"**********************************")
print("                                                                                                           ")
orig_bands_hash=generate_bands_hash(orig_sample_1000signmatrix,b,r,B)
data_bands_hash=generate_bands_hash(data_sample_1000signmatrix,b,r,B)
t=0.1
generate_cand_fp_fn(orig_bands_hash,data_bands_hash,t)

                                                                                                             
******************** Results for b = 50 r = 20 B = 199 **********************************
                                                                                                           
************************* Candidate Documents ********************************
['g0pC_taska.txt', 'g2pB_taska.txt', 'g1pB_taska.txt']
                         
************* Candidate documents and how many buckets in common with the original document ************
{'g0pC_taska.txt': 1, 'g2pB_taska.txt': 1, 'g1pB_taska.txt': 1}
                         
************* Jaccard Similarities of Candidate Documents *****************
g0pC_taska.txt 0.0638
g2pB_taska.txt 0.087
g1pB_taska.txt 0.087
                         
                                              
****************** All documents with Jaccard Similarity >t=0.1************************
['g4pB_taska.txt', 'g1pD_taska.txt',

### Optimizing with different number of bands, rows and buckets

In [281]:
# Experiment with different values of b,r,B, and t
    # to reduce the number of false positives and false negatives
    # Report all results in a table in a separate word document

In [277]:
b = 40
r = 25
B = 199
print("******************** Results for b =", str(b),"r =", str(r), "B =", str(B),"**********************************")
print("                                                                                                           ")
orig_bands_hash=generate_bands_hash(orig_sample_1000signmatrix,b,r,B)
data_bands_hash=generate_bands_hash(data_sample_1000signmatrix,b,r,B)
t=0.1
generate_cand_fp_fn(orig_bands_hash,data_bands_hash,t)

******************** Results for b = 40 r = 25 B = 199 **********************************
                                                                                                           
************************* Candidate Documents ********************************
['g4pB_taska.txt', 'g1pB_taska.txt', 'g4pE_taska.txt']
                         
************* Candidate documents and how many buckets in common with the original document ************
{'g4pB_taska.txt': 1, 'g1pB_taska.txt': 1, 'g4pE_taska.txt': 1}
                         
************* Jaccard Similarities of Candidate Documents *****************
g4pB_taska.txt 0.0667
g1pB_taska.txt 0.1111
g4pE_taska.txt 0.1111
                         
                                              
****************** All documents with Jaccard Similarity >t=0.1************************
['g0pD_taska.txt', 'g2pB_taska.txt', 'g2pE_taska.txt', 'g1pB_taska.txt', 'g4pE_taska.txt', 'g3pB_taska.txt']
                                   

In [279]:
b = 10
r = 100
B = 99
print("******************** Results for b =", str(b),"r =", str(r), "B =", str(B),"**********************************")
print("                                                                                                           ")
orig_bands_hash=generate_bands_hash(orig_sample_1000signmatrix,b,r,B)
data_bands_hash=generate_bands_hash(data_sample_1000signmatrix,b,r,B)
t=0.1
generate_cand_fp_fn(orig_bands_hash,data_bands_hash,t)

******************** Results for b = 10 r = 100 B = 99 **********************************
                                                                                                           
************************* Candidate Documents ********************************
['g0pC_taska.txt']
                         
************* Candidate documents and how many buckets in common with the original document ************
{'g0pC_taska.txt': 1}
                         
************* Jaccard Similarities of Candidate Documents *****************
g0pC_taska.txt 0.0526
                         
                                              
****************** All documents with Jaccard Similarity >t=0.1************************
['g0pD_taska.txt', 'g0pE_taska.txt', 'g0pA_taska.txt']
                                     
Count of False Positives: 1
********* False positive documents and their corresponding Jaccard Similarity *************
{'g0pC_taska.txt': 0.0526}
                         


In [280]:
b = 125
r = 8
B = 300
print("******************** Results for b =", str(b),"r =", str(r), "B =", str(B),"**********************************")
print("                                                                                                           ")
orig_bands_hash=generate_bands_hash(orig_sample_1000signmatrix,b,r,B)
data_bands_hash=generate_bands_hash(data_sample_1000signmatrix,b,r,B)
t=0.1
generate_cand_fp_fn(orig_bands_hash,data_bands_hash,t)

******************** Results for b = 125 r = 8 B = 300 **********************************
                                                                                                           
************************* Candidate Documents ********************************
['g4pB_taska.txt', 'g0pD_taska.txt', 'g1pA_taska.txt', 'g3pC_taska.txt', 'g0pA_taska.txt', 'g4pD_taska.txt', 'g3pB_taska.txt']
                         
************* Candidate documents and how many buckets in common with the original document ************
{'g4pB_taska.txt': 1, 'g0pD_taska.txt': 1, 'g1pA_taska.txt': 1, 'g3pC_taska.txt': 1, 'g0pA_taska.txt': 1, 'g4pD_taska.txt': 2, 'g3pB_taska.txt': 1}
                         
************* Jaccard Similarities of Candidate Documents *****************
g4pB_taska.txt 0.1364
g0pD_taska.txt 0.1364
g1pA_taska.txt 0.1628
g3pC_taska.txt 0.1468
g0pA_taska.txt 0.1312
g4pD_taska.txt 0.1792
g3pB_taska.txt 0.1211
                         
                                   

In [288]:
b = 25
r = 40
B = 300
print("******************** Results for b =", str(b),"r =", str(r), "B =", str(B),"**********************************")
print("                                                                                                           ")
orig_bands_hash=generate_bands_hash(orig_sample_1000signmatrix,b,r,B)
data_bands_hash=generate_bands_hash(data_sample_1000signmatrix,b,r,B)
t=0.1
generate_cand_fp_fn(orig_bands_hash,data_bands_hash,t)

******************** Results for b = 25 r = 40 B = 300 **********************************
                                                                                                           
************************* Candidate Documents ********************************
['g4pD_taska.txt']
                         
************* Candidate documents and how many buckets in common with the original document ************
{'g4pD_taska.txt': 1}
                         
************* Jaccard Similarities of Candidate Documents *****************
g4pD_taska.txt 0.0204
                         
                                              
****************** All documents with Jaccard Similarity >t=0.1************************
[]
                                     
Count of False Positives: 1
********* False positive documents and their corresponding Jaccard Similarity *************
{'g4pD_taska.txt': 0.0204}
                         
Count of False Negatives: 0
************ False negat

In [291]:
b = 100
r = 10
B = 1000
print("******************** Results for b =", str(b),"r =", str(r), "B =", str(B),"**********************************")
print("                                                                                                           ")
orig_bands_hash=generate_bands_hash(orig_sample_1000signmatrix,b,r,B)
data_bands_hash=generate_bands_hash(data_sample_1000signmatrix,b,r,B)
t=0.1
generate_cand_fp_fn(orig_bands_hash,data_bands_hash,t)

******************** Results for b = 100 r = 10 B = 1000 **********************************
                                                                                                           
************************* Candidate Documents ********************************
['g2pC_taska.txt', 'g4pD_taska.txt', 'g1pB_taska.txt', 'g4pE_taska.txt']
                         
************* Candidate documents and how many buckets in common with the original document ************
{'g2pC_taska.txt': 1, 'g4pD_taska.txt': 1, 'g1pB_taska.txt': 1, 'g4pE_taska.txt': 1}
                         
************* Jaccard Similarities of Candidate Documents *****************
g2pC_taska.txt 0.0471
g4pD_taska.txt 0.0526
g1pB_taska.txt 0.0471
g4pE_taska.txt 0.0417
                         
                                              
****************** All documents with Jaccard Similarity >t=0.1************************
[]
                                     
Count of False Positives: 4
********* Fa

#**************************************************************************************